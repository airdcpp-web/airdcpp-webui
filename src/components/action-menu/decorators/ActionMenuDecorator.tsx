import * as React from 'react';

import * as UI from 'types/ui';

import {
  ActionClickHandler,
  ActionHandlerDecorator,
} from 'decorators/ActionHandlerDecorator';
import {
  MenuFormDialog,
  MenuFormDialogProps,
} from 'components/action-menu/MenuFormDialog';
import {
  ActionMenuDefinition,
  useActionMenuItems,
} from 'components/action-menu/effects/useActionMenuItems';
import { EmptyDropdownContent } from 'components/semantic/EmptyDropdown';
import { OnShowRemoteMenuForm, useRemoteMenuItems } from '../effects/useRemoteMenuItems';

export interface ActionMenuDecoratorProps<
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
> extends ActionMenuDefinition<ItemDataT, EntityT> {
  remoteMenuNestingThreshold?: number;
  className?: string;
  caption?: React.ReactNode;
  button?: boolean;
  children?: React.ReactElement<UI.ActionMenuData<any, any>> | false;
}

interface ActionMenuNestedProps<
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
> {
  props: ActionMenuDecoratorProps<ItemDataT, EntityT>;
  onShowForm: OnShowRemoteMenuForm;
  onClickAction: ActionClickHandler<ItemDataT, EntityT>;
  onClickItem: (() => void) | undefined;
  menuBuilder: UI.ActionMenuComponentBuilder;
}

const ActionMenuNested = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>({
  props,
  onShowForm,
  onClickAction,
  onClickItem,
  menuBuilder,
}: ActionMenuNestedProps<ItemDataT, EntityT>) => {
  const { getMenuDefinitionArray, getMenuItems } = useActionMenuItems<ItemDataT, EntityT>(
    props,
  );

  const definitionArray = getMenuDefinitionArray();
  const { getRemoteItems } = useRemoteMenuItems({
    definitionArray,
    onShowForm,
    nestingThreshold: props.remoteMenuNestingThreshold,
    onClickMenuItem: onClickItem,
  });

  const getAllMenuItems = () => {
    const remoteItems = getRemoteItems();
    const items = getMenuItems(onClickAction, remoteItems, onClickItem);
    if (!items.length) {
      return <EmptyDropdownContent loading={!remoteItems} />;
    }

    return <>{menuBuilder(items)}</>;
  };

  return getAllMenuItems();
};

/*const ActionMenuNested2 = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>({
  props,
  onClickAction,
  onClickItem,
  menuBuilder,
}: ActionMenuNestedProps<ItemDataT, EntityT>) => {
  const { getMenuItems } = useActionMenuItems<ItemDataT, EntityT>(props);

  const items = getMenuItems(onClickAction, null, onClickItem);
  if (!items.length) {
    return <EmptyDropdownContent />;
  }

  return <>{menuBuilder(items)}</>;
};*/

export interface ActionMenuDecoratorChildProps {
  children: (onClick?: UI.MenuItemClickHandler) => React.ReactNode;
}

export default function <
  DropdownComponentPropsT extends object,
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  Component: React.ComponentType<ActionMenuDecoratorChildProps & DropdownComponentPropsT>,
  menuBuilder: UI.ActionMenuComponentBuilder,
) {
  type Props = ActionMenuDecoratorProps<ItemDataT, EntityT> & DropdownComponentPropsT;

  const ActionMenuDecorator: React.FC<Props> = (props) => {
    const [formHandler, setFormHandler] = React.useState<MenuFormDialogProps | null>(
      null,
    );

    return (
      <>
        <ActionHandlerDecorator<ItemDataT, EntityT>>
          {({ onClickAction }) => {
            const { actions, children, itemData, remoteMenuId, entity, ...other } = props;

            return (
              <Component
                {...(other as ActionMenuDecoratorChildProps & DropdownComponentPropsT)}
              >
                {(onClickItem) => (
                  <ActionMenuNested
                    props={props}
                    onShowForm={setFormHandler}
                    onClickAction={onClickAction}
                    onClickItem={onClickItem}
                    menuBuilder={menuBuilder}
                  />
                )}
              </Component>
            );
          }}
        </ActionHandlerDecorator>
        {!!formHandler && !!formHandler.fieldDefinitions && (
          <MenuFormDialog {...formHandler} onClose={() => setFormHandler(null)} />
        )}
      </>
    );
  };

  return ActionMenuDecorator;
}
