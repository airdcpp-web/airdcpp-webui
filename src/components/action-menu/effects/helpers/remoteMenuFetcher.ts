import { useEffect, useState } from 'react';

import * as API from 'types/api';
import * as UI from 'types/ui';

import MenuConstants, { MENU_SUPPORTS } from 'constants/MenuConstants';
import NotificationActions from 'actions/NotificationActions';
import { parseActionMenuItemId, parseActionMenuItemIds } from 'utils/MenuUtils';
import { ActionMenuDefinition } from '../useActionMenuItems';
import { useSocket } from 'context/SocketContext';

export interface RemoteMenuData {
  selectedIds: UI.ActionIdType[];
  remoteMenuId: string;
  entityId: API.IdType | undefined;
}

export interface RemoteMenu {
  menuData: RemoteMenuData;
  items: API.GroupedContextMenuItem[];
}

interface ParsedDefinitionArray {
  remoteMenuIds: Array<string | undefined>;
  selectedIdsByMenu: Array<UI.ActionIdType>[];
  entityIds: Array<API.IdType | undefined>;
}

const parseDefinitionArray = (
  definitionArray: ActionMenuDefinition<any, any>[],
): ParsedDefinitionArray => {
  const selectedIdsByMenu = definitionArray
    .map((defition) => defition.itemData)
    .map((data) => parseActionMenuItemIds(data));
  const remoteMenuIds = definitionArray.map(
    (remoteMenuProps) => remoteMenuProps.remoteMenuId,
  );
  const entityIds = definitionArray
    .map((defition) => defition.entity)
    .map((data) => parseActionMenuItemId(data) as API.IdType | undefined);

  return { selectedIdsByMenu, remoteMenuIds, entityIds };
};

export const useRemoteMenuFetcher = (
  definitionArray: ActionMenuDefinition<any, any>[],
) => {
  const { selectedIdsByMenu, remoteMenuIds, entityIds } =
    parseDefinitionArray(definitionArray);

  const [menusByType, setMenusByType] = useState<
    Array<RemoteMenu | undefined> | undefined
  >(undefined);

  const socket = useSocket();

  useEffect(() => {
    const fetchMenus = async () => {
      const menuPromises = remoteMenuIds.map(async (remoteMenuId, menuIndex) => {
        if (!remoteMenuId) {
          return Promise.resolve(undefined);
        }

        const selectedIds = selectedIdsByMenu[menuIndex];
        const entityId = entityIds[menuIndex];

        const menu = await socket.post<API.GroupedContextMenuItem[]>(
          `${MenuConstants.MODULE_URL}/${remoteMenuId}/list_grouped`,
          {
            selected_ids: selectedIds,
            entity_id: entityId,
            supports: MENU_SUPPORTS,
          },
        );

        return {
          items: menu,
          menuData: {
            remoteMenuId,
            selectedIds,
            entityId,
          },
        };
      });

      let fetchedMenus: Array<RemoteMenu | undefined>;
      try {
        fetchedMenus = await Promise.all(menuPromises);
      } catch (e) {
        NotificationActions.apiError('Failed fetch menu items', e);
        return;
      }

      setMenusByType(fetchedMenus);
    };

    fetchMenus();
  }, []);

  return menusByType;
};
