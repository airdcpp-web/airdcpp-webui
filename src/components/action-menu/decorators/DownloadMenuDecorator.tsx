import * as React from 'react';

import classNames from 'classnames';

import * as UI from 'types/ui';
import { ActionMenuDecoratorProps } from './ActionMenuDecorator';
import { DownloadableItemActionMenu } from 'actions/ui/downloadable-item';

export interface DownloadMenuDecoratorProps<
  ItemDataT extends UI.DownloadableItemInfo,
  EntityT extends UI.SessionItemBase,
> extends Omit<
      ActionMenuDecoratorProps<ItemDataT, EntityT>,
      'actions' | 'itemData' | 'caption'
    >,
    Pick<UI.DownloadableItemData<ItemDataT>, 'entity' | 'user'> {
  itemInfoGetter: () => ItemDataT;
  entity: EntityT;
  downloadHandler: UI.DownloadHandler<ItemDataT>;
  caption: React.ReactNode;
  className?: string;
}

type DownloadMenuDecoratorChildProps<
  ItemDataT extends UI.DownloadableItemInfo,
  EntityT extends UI.SessionItemBase,
> = ActionMenuDecoratorProps<UI.DownloadableItemData<ItemDataT>, EntityT>;

export default function <
  DropdownPropsT,
  ItemDataT extends UI.DownloadableItemInfo,
  EntityT extends UI.SessionItemBase,
>(
  Component: React.ComponentType<
    DownloadMenuDecoratorChildProps<ItemDataT, EntityT> & DropdownPropsT
  >,
) {
  class DownloadMenu extends React.PureComponent<
    DownloadMenuDecoratorProps<ItemDataT, EntityT> & DropdownPropsT
  > {
    itemData: UI.DownloadableItemData<ItemDataT>;
    constructor(props: DownloadMenuDecoratorProps<ItemDataT, EntityT> & DropdownPropsT) {
      super(props);

      this.itemData = {
        handler: props.downloadHandler,
      } as UI.DownloadableItemData<ItemDataT>;

      // Since table cells are recycled, the same menu can be re-used for different items
      // as it's not necessarily re-rendered due to performance reasons
      // Use getters so that we get data for the current cell
      Object.defineProperty(this.itemData, 'id', {
        get: () => {
          return this.props.itemInfoGetter().id;
        },
      });

      Object.defineProperty(this.itemData, 'entity', {
        get: () => {
          return this.props.entity;
        },
      });

      Object.defineProperty(this.itemData, 'user', {
        get: () => {
          return this.props.user;
        },
      });

      Object.defineProperty(this.itemData, 'itemInfo', {
        get: () => {
          return this.props.itemInfoGetter();
        },
      });
    }

    itemDataGetter = () => {
      return this.itemData;
    };

    render() {
      const { caption, className } = this.props;

      return (
        <Component
          {...this.props}
          className={classNames('download', className)}
          caption={caption}
          actions={DownloadableItemActionMenu}
          itemData={this.itemDataGetter}
          // entity={entity as EntityT}
        />
      );
    }
  }

  return DownloadMenu;
}
