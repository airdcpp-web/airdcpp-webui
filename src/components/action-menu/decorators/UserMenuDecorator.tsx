import * as React from 'react';
import classNames from 'classnames';
import invariant from 'invariant';

import * as UI from '@/types/ui';

import { ActionUserData } from '@/actions/ui/user';
import { getFilePath } from '@/utils/FileUtils';

import UserIcon from '@/components/icon/UserIcon';

import Icon from '@/components/semantic/Icon';
import IconConstants from '@/constants/IconConstants';
import { ActionMenuDecoratorProps } from './ActionMenuDecorator';
import { UserActionMenu } from '@/actions/ui/user';

export type UserMenuDecoratorProps = Omit<
  ActionMenuDecoratorProps<ActionUserData, UI.ActionMenuItemEntityValueType>,
  'actions' | 'caption' | 'itemData'
> &
  React.PropsWithChildren<{
    user: UI.ActionUserType;
    directory?: string;
    userIcon?: string | boolean | null;
    text?: React.ReactNode;
    className?: string;
  }>;

type UserMenuDecoratorChildProps = ActionMenuDecoratorProps<
  ActionUserData,
  UI.ActionMenuItemEntityValueType
>;

export default function <DropdownPropsT extends object>(
  Component: React.ComponentType<UserMenuDecoratorChildProps & DropdownPropsT>,
) {
  class UserMenu extends React.PureComponent<UserMenuDecoratorProps & DropdownPropsT> {
    static readonly defaultProps: Pick<UserMenuDecoratorProps, 'directory'> = {
      directory: '/',
    };

    itemData: ActionUserData;

    constructor(props: UserMenuDecoratorProps & DropdownPropsT) {
      super(props);

      this.itemData = {} as ActionUserData;
      Object.defineProperty(this.itemData, 'id', {
        get: () => {
          return this.getUserId();
        },
      });
      Object.defineProperty(this.itemData, 'user', {
        get: () => {
          return this.props.user;
        },
      });
      Object.defineProperty(this.itemData, 'directory', {
        get: () => {
          const { directory } = this.props;
          return getFilePath(directory as any as string);
        },
      });
    }

    getUserId = () => {
      const { user } = this.props;

      // User/online user
      if (!!user.id) {
        return user.id;
      }

      // Hinted user
      if (!!user.hub_url) {
        return {
          cid: user.cid,
          hub_url: user.hub_url,
        };
      }

      invariant(
        false,
        'Invalid user object in UserMenuDecorator: id and hub_url missing',
      );
    };

    itemDataGetter = () => {
      return this.itemData;
    };

    render() {
      const { text, userIcon, user, className } = this.props;

      let nicks: React.ReactNode = text;
      if (!nicks) {
        nicks = user.nicks ? user.nicks : user.nick;
      }

      let caption = nicks;
      if (userIcon) {
        caption = (
          <div className={classNames('icon-caption', userIcon)}>
            {userIcon === 'simple' ? (
              <Icon icon={IconConstants.USER} color="blue" />
            ) : (
              <UserIcon size="large" flags={user.flags} />
            )}
            {nicks}
          </div>
        );
      }

      return (
        <Component
          {...this.props}
          className={classNames('user-menu', className)}
          caption={caption}
          actions={UserActionMenu}
          itemData={this.itemDataGetter}
        />
      );
    }
  }

  return UserMenu;
}
