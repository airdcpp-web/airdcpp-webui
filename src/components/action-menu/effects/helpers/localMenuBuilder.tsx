import { toActionI18nKey } from '@/utils/ActionUtils';

import * as UI from '@/types/ui';

import { ActionClickHandler } from '@/decorators/ActionHandlerDecorator';
import { Trans } from 'react-i18next';
import { parseTranslationModules } from '@/utils/TranslationUtils';

const getDivider = (menuIndex: number, itemIndex: number) => {
  return {
    id: `divider${menuIndex}_${itemIndex}`,
  };
};

const actionToActionMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  action: UI.MenuActionDefition<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
) => {
  const active = !action.checked ? false : action.checked(menu.itemDataGetter());
  const icon = !!action.checked ? (active ? 'checkmark' : '') : action.icon;
  return {
    id: action.id,
    item: {
      onClick: () => {
        onClickAction({
          action,
          itemData: menu.itemDataGetter(),
          entity: menu.entity,
          moduleData: menu.moduleData,
        });
      },
      active,
      icon,
      children: (
        <Trans
          i18nKey={toActionI18nKey(
            action,
            parseTranslationModules(menu.moduleData.moduleId),
          )}
          defaults={action.displayName}
        >
          {action.displayName}
        </Trans>
      ),
    },
  };
};

// Convert ID to menu link element
const getMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
  action: UI.MenuActionListItemType<ItemDataT, EntityT>,
  itemIndex: number,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
): UI.ActionMenuItem => {
  if (!action) {
    return getDivider(menuIndex, itemIndex);
  }

  if ('children' in action) {
    const children = action.children.map((childAction, childActionIndex) => {
      if (!childAction) {
        return getDivider(0, childActionIndex);
      }

      return actionToActionMenuItem(childAction, menu, onClickAction);
    });

    if (children.length === 1) {
      return children[0];
    }

    return {
      id: action.id,
      item: {
        icon: action.icon,
        onClick: () => {
          // ..
        },
        children: <>{action.displayName}</>,
      },
      children,
    };
  }

  return actionToActionMenuItem(action, menu, onClickAction);
};

export const localMenuToActionMenuItems = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
) => {
  const items = menu.actions.map((actionId, actionIndex) => {
    return getMenuItem(
      menu as UI.ActionMenuType<ItemDataT, EntityT>,
      menuIndex,
      actionId,
      actionIndex,
      onClickAction,
    );
  });

  return items;
};
