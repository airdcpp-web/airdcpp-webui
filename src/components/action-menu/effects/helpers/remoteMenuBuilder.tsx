import * as API from '@/types/api';
import * as UI from '@/types/ui';

import MenuConstants from '@/constants/MenuConstants';
import IconConstants from '@/constants/IconConstants';
import { RemoteMenu, RemoteMenuData } from './remoteMenuFetcher';
import { UILogger } from '@/effects/LoggerEffect';

export interface CombinedContextMenuItem {
  typeData: RemoteMenuData;
  item: API.ContextMenuItem;
}

export interface CombinedContextMenu extends Omit<API.GroupedContextMenu, 'items'> {
  items: CombinedContextMenuItem[];
}

interface MenuOptions {
  handleSelect: SelectHandler;
  nestingThreshold: number;
}

type SelectHandler = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
) => void;

const toMenuItem = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
  handleSelect: SelectHandler,
): UI.ActionMenuItem => {
  const onClick = () => {
    handleSelect(menuItem, remoteMenuData);
  };

  const { id, icon, title } = menuItem;
  return {
    id,
    item: {
      onClick,
      icon: icon[MenuConstants.ICON_TYPE_ID],
      children: title,
    },
  };
};

export const remoteMenuToActionMenuItems = (
  menu: CombinedContextMenu,
  { handleSelect, nestingThreshold }: MenuOptions,
): UI.ActionMenuItem[] => {
  const items = menu.items.map(({ item, typeData }) =>
    toMenuItem(item, typeData, handleSelect),
  );

  if (items.length <= nestingThreshold) {
    return items;
  }

  return [
    {
      id: menu.id,
      item: {
        icon: menu.icon[MenuConstants.ICON_TYPE_ID] || IconConstants.EXTENSION,
        onClick: () => {
          // ..
        },
        children: <>{menu.title}</>,
      },
      children: items,
    },
  ];
};

// Combine menus by their subscriber ID and remove duplicate items
export const createRemoteMenuCombineReducer = (logger: UILogger) => {
  return (reduced: CombinedContextMenu[], typeMenus: RemoteMenu | undefined) => {
    if (!typeMenus) {
      return reduced;
    }

    for (const menu of typeMenus.menus) {
      const existing = reduced.find((m) => m.id === menu.id);

      const toMenuItem = (item: API.ContextMenuItem): CombinedContextMenuItem => ({
        typeData: typeMenus.typeData,
        item,
      });

      if (existing) {
        existing.items.push(
          ...menu.items
            .filter((item) => {
              const existingItem = existing.items.find(
                (existingItem) =>
                  existingItem.item.id === item.id &&
                  existingItem.item.hook_id === item.hook_id,
              );
              if (existingItem) {
                logger.info(
                  // eslint-disable-next-line max-len
                  `Removing a duplicate item ${item.id} (subscriber ${menu.id}) from menu ${typeMenus.typeData.remoteMenuId} (exists for menu ${existingItem.typeData.remoteMenuId})`,
                );
              }

              return !existingItem;
            })
            .map(toMenuItem),
        );
      } else {
        reduced.push({
          ...menu,
          items: menu.items.map(toMenuItem),
        });
      }
    }

    return reduced;
  };
};
