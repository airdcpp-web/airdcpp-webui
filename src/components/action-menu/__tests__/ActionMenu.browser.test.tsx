import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { renderDataRoutes } from '@/tests/render/test-renderers';

import { waitFor } from '@testing-library/dom';

import {
  clickButton,
  waitExpectRequestToMatchSnapshot,
} from '@/tests/helpers/test-helpers';
import { setInputFieldValues, setupUserEvent } from '@/tests/helpers/test-form-helpers';
import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import {
  RemoteMenuGrouped1,
  RemoteMenu1ItemForm,
  RemoteMenu1ItemNormal,
  RemoteMenuGrouped2,
  RemoteMenu1ItemWithChildren,
} from '@/tests/mocks/api/menu';
import {
  ActionMenu as ActionDropdownMenu,
  TableActionMenu,
} from '@/components/action-menu';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';
import {
  getTestActionMenu,
  installActionMenuMocks,
  TestActionItemType,
  TestItemEnableCaption,
  TestItemRemoveApproveCaption,
  TestItemRemoveCaption,
} from './test-action-menu-helpers';
import {
  clickMenuItem,
  clickSubMenu,
  openMenu,
  waitMenuClosed,
} from '@/tests/helpers/test-menu-helpers';
import { RemoteMenuData } from '../effects/helpers/remoteMenuFetcher';
import { ActionMenuDecoratorProps } from '../decorators/ActionMenuDecorator';

const RemoteMenuId = 'test-remote-menu-id';

const TriggerCaption = 'Open menu';
// tslint:disable:no-empty
describe('Action menu', () => {
  let server: MockServer;

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);

    // Remote menus
    const menuMocks = installActionMenuMocks(
      RemoteMenuId,
      [RemoteMenuGrouped1, RemoteMenuGrouped2],
      server,
    );

    const userEvent = setupUserEvent();
    return { commonData, server, menuMocks, userEvent };
  };

  type ActionMenuComponentType = React.ComponentType<
    ActionMenuDecoratorProps<TestActionItemType, TestActionItemType> &
      Partial<RemoteMenuData>
  >;

  const renderMenu = async (
    Component: ActionMenuComponentType,
    useRemoteMenu = false,
  ) => {
    const { commonData, server, ...other } = await getSocket();

    const { menu, ...actionHandlers } = getTestActionMenu();

    const onSave = vi.fn(() => Promise.resolve());

    const ActionDropdownMenuTest = () => {
      const testItemData = { id: 'test', name: 'Test Item' };
      const testEntity = { id: 'test-entity', name: 'Test Entity' };
      return (
        <Component
          caption={TriggerCaption}
          button={true}
          actions={menu}
          itemData={testItemData}
          entity={testEntity}
          remoteMenuId={useRemoteMenu ? RemoteMenuId : undefined}
          remoteMenuNestingThreshold={2}
        />
      );
    };

    const routes = [
      {
        index: true,
        Component: ActionDropdownMenuTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/'] },
      viewType: VIEW_FIXED_HEIGHT,
    });

    return { onSave, ...renderData, ...other, ...actionHandlers };
  };

  const createMenuTests = (Component: ActionMenuComponentType) => {
    describe('Local actions', () => {
      test('should select menu item', async () => {
        const renderResult = await renderMenu(Component);
        const { onEnable } = renderResult;

        await openMenu(TriggerCaption, renderResult);
        await clickMenuItem(TestItemEnableCaption, renderResult);

        await waitFor(() => expect(onEnable).toHaveBeenCalledTimes(1));
        await waitMenuClosed(renderResult);
      }, 100000);

      test('should handle confirm data', async () => {
        const renderResult = await renderMenu(Component);
        const { onRemove, getByText, getByRole } = renderResult;

        await openMenu(TriggerCaption, renderResult);
        await clickMenuItem(TestItemRemoveCaption, renderResult);

        await waitFor(() => expect(getByText(TestItemRemoveApproveCaption)).toBeTruthy());
        expect(onRemove).toHaveBeenCalledTimes(0);

        await clickButton(TestItemRemoveApproveCaption, getByRole);
        await waitFor(() => expect(onRemove).toHaveBeenCalledTimes(1));
        await waitMenuClosed(renderResult);
      }, 100000);
    });

    describe('Remote actions', () => {
      test('should select menu item', async () => {
        const renderResult = await renderMenu(Component, true);
        const { menuMocks } = renderResult;

        await openMenu(TriggerCaption, renderResult);

        await waitFor(() => expect(menuMocks.onListGrouped).toHaveBeenCalledTimes(1));

        const { clickSubMenuItem } = await clickSubMenu(
          RemoteMenuGrouped1.title,
          renderResult,
        );
        await clickSubMenuItem(RemoteMenu1ItemNormal.title);

        await waitExpectRequestToMatchSnapshot(menuMocks.onSelectMenuItem);
        await waitMenuClosed(renderResult);
      }, 100000);

      test('should handle forms', async () => {
        const renderResult = await renderMenu(Component, true);
        const { menuMocks, getByRole } = renderResult;

        await openMenu(TriggerCaption, renderResult);

        await waitFor(() => expect(menuMocks.onListGrouped).toHaveBeenCalledTimes(1));

        const { clickSubMenuItem } = await clickSubMenu(
          RemoteMenuGrouped1.title,
          renderResult,
        );
        await clickSubMenuItem(RemoteMenu1ItemForm.title);

        await waitFor(() => expect(getByRole('dialog')).toBeTruthy());

        await setInputFieldValues(renderResult, {
          'Text to show': 'My text',
        });

        await clickButton('Continue', getByRole);

        await waitExpectRequestToMatchSnapshot(menuMocks.onSelectMenuItem);
        await waitMenuClosed(renderResult);
      }, 100000);

      test('should handle deep nesting', async () => {
        const renderResult = await renderMenu(Component, true);
        const { menuMocks } = renderResult;

        await openMenu(TriggerCaption, renderResult);

        await waitFor(() => expect(menuMocks.onListGrouped).toHaveBeenCalledTimes(1));

        // Go to first level
        const { clickSubMenu: clickSubMenuLevel1 } = await clickSubMenu(
          RemoteMenuGrouped1.title,
          renderResult,
        );

        // Go to second level
        const level1Item = RemoteMenu1ItemWithChildren;
        const { clickSubMenu: clickSubMenuLevel2 } = await clickSubMenuLevel1(
          level1Item.title,
        );

        // Go to third level
        const level2Item = level1Item.children[0];
        const { clickSubMenuItem } = await clickSubMenuLevel2(level2Item.title);

        // TODO: menus use different role for the return button
        // Test return to parent
        //await returnToParent();

        // Go back to third level
        //const level3Item = level2Item.children![0];
        //const { clickSubMenuItem } = await clickSubMenuLevel2(level2Item.title);

        // Click an item
        const level3Item = level2Item.children![0];
        await clickSubMenuItem(level3Item.title);

        await waitExpectRequestToMatchSnapshot(menuMocks.onSelectMenuItem);
      }, 100000);
    });
  };

  describe('ActionDropdownMenu', () => {
    createMenuTests(ActionDropdownMenu);
  });

  describe('TableActionMenu', () => {
    createMenuTests(TableActionMenu);
  });
});
