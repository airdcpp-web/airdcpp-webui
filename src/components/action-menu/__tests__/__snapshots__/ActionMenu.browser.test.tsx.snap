// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Action menu > ActionDropdownMenu > Remote actions > should handle forms 1`] = `
{
  "data": {
    "entity_id": "test-entity",
    "form_definitions": [
      {
        "default_value": false,
        "key": "boolean",
        "title": "Yes/no",
        "type": "boolean",
      },
      {
        "default_value": "",
        "key": "string",
        "title": "Text to show",
        "type": "string",
      },
      {
        "default_value": "",
        "key": "hub_url",
        "optional": true,
        "title": "Hub URL",
        "type": "hub_url",
      },
      {
        "default_value": "",
        "key": "email",
        "optional": true,
        "title": "Email",
        "type": "email",
      },
      {
        "default_value": null,
        "key": "hinted_user",
        "optional": true,
        "title": "Hinted User",
        "type": "hinted_user",
      },
      {
        "default_value": "",
        "key": "password",
        "optional": true,
        "title": "Password",
        "type": "password",
      },
      {
        "default_value": "",
        "key": "file_path",
        "optional": true,
        "title": "File path",
        "type": "file_path",
      },
      {
        "default_value": "",
        "key": "directory_path",
        "optional": true,
        "title": "Directory path",
        "type": "directory_path",
      },
      {
        "default_value": "",
        "key": "existing_file_path",
        "optional": true,
        "title": "Existing file path",
        "type": "existing_file_path",
      },
      {
        "default_value": "",
        "key": "text",
        "optional": true,
        "title": "Text",
        "type": "text",
      },
      {
        "default_value": [],
        "definitions": [
          {
            "default_value": 1,
            "key": "number",
            "min": 0,
            "title": "Number",
            "type": "number",
          },
          {
            "default_value": "",
            "key": "url",
            "optional": true,
            "title": "URL",
            "type": "url",
          },
        ],
        "item_type": "struct",
        "key": "list",
        "optional": true,
        "title": "List",
        "type": "list",
      },
    ],
    "form_value": {
      "boolean": false,
      "directory_path": null,
      "email": null,
      "existing_file_path": null,
      "file_path": null,
      "hinted_user": null,
      "hub_url": null,
      "list": [],
      "password": null,
      "string": "My text",
      "text": null,
    },
    "hook_id": "airdcpp-test",
    "menuitem_id": "test_form",
    "selected_ids": [
      "test",
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/test-remote-menu-id/select",
}
`;

exports[`Action menu > ActionDropdownMenu > Remote actions > should select menu item 1`] = `
{
  "data": {
    "entity_id": "test-entity",
    "form_definitions": null,
    "hook_id": "airdcpp-test",
    "menuitem_id": "test_item_id",
    "selected_ids": [
      "test",
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/test-remote-menu-id/select",
}
`;

exports[`Action menu > TableActionMenu > Remote actions > should handle forms 1`] = `
{
  "data": {
    "entity_id": "test-entity",
    "form_definitions": [
      {
        "default_value": false,
        "key": "boolean",
        "title": "Yes/no",
        "type": "boolean",
      },
      {
        "default_value": "",
        "key": "string",
        "title": "Text to show",
        "type": "string",
      },
      {
        "default_value": "",
        "key": "hub_url",
        "optional": true,
        "title": "Hub URL",
        "type": "hub_url",
      },
      {
        "default_value": "",
        "key": "email",
        "optional": true,
        "title": "Email",
        "type": "email",
      },
      {
        "default_value": null,
        "key": "hinted_user",
        "optional": true,
        "title": "Hinted User",
        "type": "hinted_user",
      },
      {
        "default_value": "",
        "key": "password",
        "optional": true,
        "title": "Password",
        "type": "password",
      },
      {
        "default_value": "",
        "key": "file_path",
        "optional": true,
        "title": "File path",
        "type": "file_path",
      },
      {
        "default_value": "",
        "key": "directory_path",
        "optional": true,
        "title": "Directory path",
        "type": "directory_path",
      },
      {
        "default_value": "",
        "key": "existing_file_path",
        "optional": true,
        "title": "Existing file path",
        "type": "existing_file_path",
      },
      {
        "default_value": "",
        "key": "text",
        "optional": true,
        "title": "Text",
        "type": "text",
      },
      {
        "default_value": [],
        "definitions": [
          {
            "default_value": 1,
            "key": "number",
            "min": 0,
            "title": "Number",
            "type": "number",
          },
          {
            "default_value": "",
            "key": "url",
            "optional": true,
            "title": "URL",
            "type": "url",
          },
        ],
        "item_type": "struct",
        "key": "list",
        "optional": true,
        "title": "List",
        "type": "list",
      },
    ],
    "form_value": {
      "boolean": false,
      "directory_path": null,
      "email": null,
      "existing_file_path": null,
      "file_path": null,
      "hinted_user": null,
      "hub_url": null,
      "list": [],
      "password": null,
      "string": "My text",
      "text": null,
    },
    "hook_id": "airdcpp-test",
    "menuitem_id": "test_form",
    "selected_ids": [
      "test",
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/test-remote-menu-id/select",
}
`;

exports[`Action menu > TableActionMenu > Remote actions > should select menu item 1`] = `
{
  "data": {
    "entity_id": "test-entity",
    "form_definitions": null,
    "hook_id": "airdcpp-test",
    "menuitem_id": "test_item_id",
    "selected_ids": [
      "test",
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/test-remote-menu-id/select",
}
`;
